geometry/symmetric_bilinear_form_body.hpp

```diff
@@ -42,6 +55,64 @@ SymmetricBilinearForm<Scalar, Frame>::operator()(
   return InnerProduct(left, *this * right);
 }
 
+template<typename Scalar, typename Frame>
+template<typename Eigenframe>
+typename SymmetricBilinearForm<Scalar, Frame>::Eigensystem<Eigenframe>
+SymmetricBilinearForm<Scalar, Frame>::Diagonalize() const {
+  R3x3Matrix<Scalar> const& A = matrix_;
+  auto const I = R3x3Matrix<double>::Identity();
+
+  // This algorithm follows
+  // https://en.wikipedia.org/wiki/Eigenvalue_algorithm#3%C3%973_matrices which
+  // gives closed-form formulæ for 3x3 matrices.
+  Scalar const q = A.Trace() / 3;
+  R3x3Matrix<Scalar> const A_minus_qI = A - q * I;
+  Scalar const p = Sqrt((A_minus_qI * A_minus_qI).Trace() / 6);
+  R3x3Matrix<double> const B = A_minus_qI / p;
+  double const det_B = B.Determinant();
+  Angle const θ = ArcCos(det_B * 0.5);
+  double const β₀ = 2 * Cos(θ / 3);
+  double const β₁ = 2 * Cos((θ + 2 * π * Radian) / 3);
+  double const β₂ = 2 * Cos((θ + 4 * π * Radian) / 3);
+  std::array<Scalar, 3> αs = {p * β₀ + q, p * β₁ + q, p * β₂ + q};
+  std::sort(αs.begin(), αs.end());
+  Scalar const& α₀ = αs[0];
+  Scalar const& α₁ = αs[1];
+  Scalar const& α₂ = αs[2];
+
+  // The form in its diagonal basis.
+  Scalar const zero;
+  SymmetricBilinearForm<Scalar, Eigenframe> form(
+      R3x3Matrix<Scalar>({  α₀, zero, zero},
+                         {zero,   α₁, zero},
+                         {zero, zero,   α₂}));
+
+  // Use the Cayley-Hamilton theorem to efficiently find the eigenvectors.  The
+  // m matrices contain, in columns, eigenvectors for the corresponding α.
+  // However it's possible for a column to be identically 0.  To deal with this
+  // the call to PickEigenvector extracts the column with the largest norm.
+  R3x3Matrix<Scalar> const A_minus_α₀I = A - α₀ * I;
+  R3x3Matrix<Scalar> const A_minus_α₁I = A - α₁ * I;
+  R3x3Matrix<Scalar> const A_minus_α₂I = A - α₂ * I;
+  auto const m₀ = A_minus_α₁I * A_minus_α₂I;
+  auto const m₁ = A_minus_α₂I * A_minus_α₀I;
+  auto const m₂ = A_minus_α₀I * A_minus_α₁I;
+  auto const v₀ = PickEigenvector(m₀);
+  auto const v₁ = PickEigenvector(m₁);
+
+  // The vectors (v₀, v₁, v₂) forms an orthonormal basis.  Make sure that it is
+  // direct.
+  auto v₂ = PickEigenvector(m₂);
+  if (R3x3Matrix(v₀, v₁, v₂).Determinant() < 0) {
+    v₂ = -v₂;
+  }
+  Rotation<Frame, Eigenframe> const rotation{
+      Vector<double, Frame>(v₀),
+      Vector<double, Frame>(v₁),
+      Bivector<double, Frame>(v₂)};  // This bivector is a bit dodgy.
```

Comment:
The bivector is not dodgy, the if statement above where we manually mess with the orientation is; the third argument here should just be `Wedge(v₀, v₁)`, we do not need `v₂`.

Reply:
Done.